pkg/chezmoi/archive.go

```diff
@@ -145,24 +181,28 @@ HEADER:
 		case err != nil:
 			return err
 		}
-		for _, skippedDirPrefix := range skippedDirPrefixes {
-			if strings.HasPrefix(header.Name, skippedDirPrefix) {
-				continue HEADER
-			}
-		}
-		name := strings.TrimSuffix(header.Name, "/")
 		switch header.Typeflag {
-		case tar.TypeDir, tar.TypeReg:
-			switch err := f(name, header.FileInfo(), tarReader, ""); {
-			case errors.Is(err, fs.SkipDir):
-				skippedDirPrefixes = append(skippedDirPrefixes, header.Name)
-			case errors.Is(err, Break):
-				return nil
-			case err != nil:
-				return err
+		case tar.TypeReg, tar.TypeDir, tar.TypeSymlink:
+			if header.Typeflag == tar.TypeReg {
+				dirs, _ := filepath.Split(header.Name)
+				dirComponents := strings.Split(strings.TrimSuffix(dirs, "/"), "/")
```

Comment:
The path manipulation functions here shout use the [`path` package](https://pkg.go.dev/path) instead of `filepath` and `strings`. `filepath` has OS-specific path handling code (e.g. it handles `\` as a path separator on Windows, but not on Linux). Since we're reading from a tar archive here, we can assume `/`-separated paths, i.e. use the `path` package.